src/ucan/internal.ts

```diff
@@ -0,0 +1,123 @@
+import localforage from 'localforage'
+
+import * as common from '../common'
+import * as pathUtil from '../fs/path'
+import * as ucan from '../ucan'
+import { UCANS_STORAGE_KEY } from '../common'
+import { Prerequisites } from './prerequisites'
+import { Ucan, WNFS_PREFIX } from '../ucan'
+import { setup } from '../setup/internal'
+
+
+let dictionary: Record<string, Ucan> = {}
+
+
+// FUNCTIONS
+
+
+/**
+ * You didn't see anything 👀
+ */
+export async function clearStorage(): Promise<void> {
+  dictionary = {}
+  await localforage.removeItem(UCANS_STORAGE_KEY)
+}
+
+/**
+ * Lookup the prefix for a filesystem key in the dictionary.
+ */
+export function dictionaryFilesystemPrefix(username: string): string {
+  // const host = `${username}.${setup.endpoints.user}`
+  // TODO: Waiting on API change.
+  //       Should be `${WNFS_PREFIX}:${host}/`
+  return WNFS_PREFIX + ":/"
+}
+
+/**
+ * Look up a UCAN with a file system path.
+ */
+export async function lookupFilesystemUcan(path: string): Promise<Ucan | null> {
+  const pathParts = pathUtil.splitParts(path)
+  const username = await common.authenticatedUsername()
+  const prefix = username ? dictionaryFilesystemPrefix(username) : ""
+
+  if (dictionary["*"]) {
+    return dictionary["*"]
+  }
+
+  return pathParts.reduce(
+    (acc: Ucan | null, part: string, idx: number) => {
+      if (acc) return acc
+      const partialPath = pathUtil.join(pathParts.slice(0, pathParts.length - idx))
+      return dictionary[`${prefix}${partialPath}`] || null
+    },
+    null
+  )
+}
+
+/**
+ * Store UCANs and update the in-memory dictionary.
+ */
+export async function store(ucans: Array<string>): Promise<void> {
+  const existing = await localforage.getItem(UCANS_STORAGE_KEY) as Array<string>
+  const newList = (existing || []).concat(ucans)
+
+  dictionary = ucan.compileDictionary(newList)
+
+  const filteredList = listFromDictionary()
+  const encodedList = filteredList.map(ucan.encode)
+
+  await localforage.setItem(UCANS_STORAGE_KEY, encodedList)
+}
+
+/**
+ * See if the stored UCANs in the in-memory dictionary
+ * conform to the given `Prerequisites`.
+ */
+export function validatePrerequisites(
+  { app, fs }: Prerequisites,
+  username: string
+): boolean {
+  const prefix = dictionaryFilesystemPrefix(username)
+
+  // Root access
+  const rootUcan = dictionary["*"]
+  if (rootUcan && !ucan.isExpired(rootUcan)) return true
+
+  // Check prerequisites
+  if (app) {
+    const u = dictionary[`${prefix}private/Apps/${app.creator}/${app.name}`]
```

Comment:
Nodes on the private side need to be addressed by their "namefilter" (something like `boris.fission.name/private/a0a7de4f90a....`) not their path in the filesystem.

I see two difficulties here: 
- you can't know a namefilter without having read access to it. So how do you request access? Request access for paths & then the auth lobby (which does have read access already) will translate that into namefilters for the UCAN composition?
- we need a merkle proof with every change that gets sent along on the private side

Ya know this might be worth having a slightly longer form discussion on the forum. I imagine I'll be helping out on this at least with the merkle proof side

Reply:
Actually we'll probably want to store these in the dictionary with the path as the key like you're doing here since the private names don't immediately reveal info about the scope.

It looks like the dictionary is keyed by the `rsc` on the UCAN. In the case of private nodes that `rsc` will be the name filter. 

UCANs have to discuss private nodes using their name filters, but at the sdk level, we may want to discuss private nodes using their path (either "absolute path" from the root of the filesystem, or "relative path" from the app folder)

Reply:
Alright gonna post some thoughts in Talk

Reply:
https://talk.fission.codes/t/ucans-on-the-private-side/988

Reply:
Oh right, forgot about those name filters.
Are name filters one way? (ie. from path -> name filter)
Or can I get a path from a name filter as well?

Reply:
I realize this is a lot to lump on at once + namefitlers weren't even a construct in the code you were working with until I got my PR in yesterday :sweat_smile: 

I've been in the land of namefilters for the past couple of weeks so this is all obviously at the top of my mind right now. More than happy to answer any questions & give a hand!

I think the Merkle proofs are out of the scope of this PR (and will probably fall on me to do :thinking:)

But we probably do want to get Private UCAN's talking in terms of namefilters. Check out Brooke's & my discussion on the talk forum. The client will still pass the auth lobby the canonical paths to private folders. The auth lobby will translate those into namefilters & issue UCANs for them. The sdk will talk about them in terms of their human-readable path, but the actual `rsc` field will be a namefilter

Reply:
> Are name filters one way? (ie. from path -> name filter)
> Or can I get a path from a name filter as well?

They’re not reversible (well, it’s POSSIBLE but completely impractical). Also, the paths here are the AES keys, not the human readable names

Reply:
oops didn't see your response when I posted another reply!

yup as brooke said name filters are made from AES keys so you can path -> namefilter, but you need read access to the FS so it's more like:  `(path, key, filesystem) -> namefilter` (which is why only the auth lobby can calculate this)

to go namefilter -> path, I think you would have to walk the tree from the root until you find the node in question :thinking: so yeah pretty impractical

Reply:
> to go namefilter -> path, I think you would have to walk the tree from the root until you find the node in question 🤔 so yeah pretty impractical

wrt the above quote, but actually a question for @icidasset: is there a specific use case in mind, or just figuring out the bounds of the system?

Reply:
> But we probably do want to get Private UCAN's talking in terms of namefilters. Check out Brooke's & my discussion on the talk forum. The client will still pass the auth lobby the canonical paths to private folders. The auth lobby will translate those into namefilters & issue UCANs for them.

Ah makes sense now, thanks. Yeah, let's implement that in a different PR 👍 

> question for @icidasset: is there a specific use case in mind, or just figuring out the bounds of the system?

Was confused about what I needed to do here. But it sounds like I can implement this later in the auth lobby and then adjust the SDK accordingly.

Reply:
Yup most of this change is in the auth lobby :+1: 

I think the one change in this code is in `compileDictionary`, the dictionary is being keyed by the UCAN's `rsc`. Right now that's the private path, but it _will be_ the name filter. I think we want it to remain keyed by the private path. But yeah this probably doesn't make much sense to do until we're sending back namefilter UCANs from the auth lobby :thinking:

Reply:
Sounds good, let's do another PR 👍